pkg/util/ringq/dyn_chan.go

```diff
@@ -0,0 +1,177 @@
+package ringq
+
+import (
+	"context"
+	"errors"
+)
+
+// Package level named errors.
+var (
+	ErrDynChanClosed = errors.New("closed DynChan")
+)
+
+// DynChan is the same as DynChanMax(minBufLen, -1).
+func DynChan[T any](minBufLen int) (chan<- T, <-chan T, ChanStatsReader) {
+	return DynChanMax[T](minBufLen, -1)
+}
+
+// DynChanMax provides a simple queueing system based on a send-only, a
+// receive-only, and an internal ring buffer queue, which has dynamic capcity.
+// It also provides a ChanStatsReader to allow querying basic stats. The
+// internal ring buffer is initially empty, and its capacity grows as more items
+// need to be stored if the receiving side is lagging behind. When the receiving
+// side catches up, the buffer will shrink to reclaim unused memory. The
+// argument minBufLen must be positive, and allows customizing the minimum
+// allocation for the internal buffer. This is useful to reduce the overhead of
+// resizing the buffer within an acceptable, permanent allocation of minBufLen
+// items ready to be used. If maxBufLen is positive, then it must be greater
+// than minBufLen, and will be used to prevent buffering more than that number
+// of items. If more items are received and the buffer already has maxBufLen,
+// then the oldest item (the one that receiving would first get) will be
+// overwritten with the new value.
+func DynChanMax[T any](minBufLen, maxBufLen int) (chan<- T, <-chan T, ChanStatsReader) {
+	if minBufLen < 1 || (maxBufLen > 0 && maxBufLen < minBufLen) {
+		panic("maxBufLen > 0 && maxBufLen < minBufLen")
+	}
+
+	var stats ChanStats
+	in := make(chan T)
+	out := make(chan T)
+	statsChan := make(chan ChanStats)
+	sr := ChanStatsReader{
+		c: statsChan,
+	}
+	if maxBufLen > 0 {
+		stats.MaxCap = maxBufLen
+	}
+
+	go func() {
+		defer close(out)
+		defer close(statsChan)
+
+		var q RingQ[T]
+		q.Max = maxBufLen
+
+		// the loop condition is that we either have items to dequeue
+		// (q.Len()>0) or that we have the possibility to receive new items to
+		// be queued (in != nil)
+		for stats.Len > 0 || in != nil {
+			// if we don't have anything in the queue, then make the dequeueing
+			// branch of the select block indefinitely by providing a nil
+			// channel, leaving only the queueing branch available
+			dequeueChan := out
+			if stats.Len == 0 {
+				dequeueChan = nil
+			}
+
+			select {
+			case v, ok := <-in: // blocks until something is queued
+				if !ok {
+					// in was closed, so if we leave it like that the next
+					// iteration will keep receiving zero values with ok=false
+					// without any blocking. So we set in to nil, so that
+					// the next iteration the select will block indefinitely on
+					// this branch of the select and leave only the dequeing
+					// branch active until all items have been dequeued
+					in = nil
+				} else {
+					if stats.Cap == 0 {
```

Comment:
Do we need this case? Doesn't calling `q.Enqueue(v)` automatically grow it if needed?

Reply:
good question, if we also had q.Min then we could handle it completely within Enqueue and prevent Shrink from reducing the capacity below q.Min, then we wouldn't need this.

Reply:
Yes it does, but we want to make sure that we start off with `minBufLen` instead just a handful of elements. This is because maybe `minBufLen` is 1000 because we anticipate a lot of traffic, so we use the user provided `minBufLen` as a hint that we don't want to allocate less than that, because that would mean doing a lot of small allocations to accommodate for small changes.

`q.Enqueue` is good to keep up with growing as needed, but if for example we just let it start from scratch then we will make a first allocation for 2 items, the next allocation will be for 5 elements, the next 10, 17, 28, etc. That means a lot of small allocations and a lot of calls to `migrate` to move around items when traffic is low. That's a lot of overhead, and each allocation adds pressure to the garbage collector, so we give the user the ability to pick a sensitive `minBufLen` to start with something more reasonable, e.g. you could pick 128 (depending on how you anticipate your workload). That means we will have made a single initial allocation and as long as you don't go over 128, you just keep looping around the same slice, there is not a lot of copying from slice to slice and we don't add unnecessary pressure to the garbage collector.

As for picking "the right" `minBufLen`, use the same reasoning you would for picking your regular channel buffer when you do `make(chan T, bufLen)`, you need to make the same decision there but here the buffer can grow past that number and then shrink to recover.

Reply:
@DanCech I thought about adding `Min` to `RingQ`. `RingQ` grows automatically, so it makes sense to set a max for it. But it doesn't shrink automatically, so `Min` only makes sense for the first allocation, which can be replaced by just calling `Grow`. On the other hand, `DynChan` grows _and_ shrinks automatically, so it makes sense to allow customizing the min as well.

Reply:
On the flip side, you could also move Min into the ring buffer and the shrinking logic into Dequeue(), so that the buffer itself can dynamically modify its size, then the channel can just set the ring min and max, and wouldn't need to contain the initial grow or the shrink after dequeue code, but can still maintain the stats by checking whether q.Cap() != stats.Cap as it does after Enqueue

Reply:
I think that's a cool idea, it makes sense to have shrinking logic in RingQ as well. I also think that the current shrinking logic could use some improvements, but I think it's something to start with. I think in general I just wanted RingQ to be a wrapping slice, but I think as a general purpose ring buffer it could also implement more features, like auto shrinking, and some other methods to be able to use it as a stack as well, not just a fifo.

Reply:
A few weeks ago I started writing something similar to RingQ just for fun, with a lot of features and experimenting with the API: https://github.com/diegommm/xgo/tree/main/list
Most of the things there are overkill here, but some could be borrowed or changed.

Reply:
@DanCech made the changes to have Min in the ring, and refactored to also move stats into the ring. Renamed the package and improved the API to be more consistent. Renamed `RingQ` to just `Ring`, so if we want to add non-queue methods later we don't need to rename it. Changed `DynChan` and `DynChanMax` to just `AdaptiveChan`, which is simpler (and IMO sounds cooler). The new controller allows changing Min and Max after creation, and also to just Grow/Shrink and even Clear, the latter is meant as an administrative operation.